File: go/corpus_test.go
package hypercast

// Replays the shared conformance corpus (corpus/*.json at the repository root) — the same
// files every other binding's suite replays. Fault spans are byte offsets into the UTF-8
// input, which is exactly what these doors receive, so span assertions hold verbatim.

import (
	"encoding/json"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"testing"
	"time"

	"github.com/google/uuid"
)

type vector struct {
	Input     string          `json:"input"`
	Expect    string          `json:"expect"`
	Type      string          `json:"type"`
	Value     json.RawMessage `json:"value"`
	Format    *formatSpec     `json:"format"`
	FaultSpan []int           `json:"fault"`
	Precision uint32          `json:"precision"`
	Seconds   *int64          `json:"seconds"`
	Nanos     int64           `json:"nanos"`
	Year      *int            `json:"year"`
	Month     int             `json:"month"`
	Day       int             `json:"day"`
}

type formatSpec struct {
	DecimalSep string `json:"decimal_sep"`
	GroupSep   string `json:"group_sep"`
	Flags      uint32 `json:"flags"`
}

func corpus(t *testing.T, name string) []vector {
	t.Helper()
	dir, err := os.Getwd()
	if err != nil {
		t.Fatal(err)
	}
	for ; dir != "/"; dir = filepath.Dir(dir) {
		candidate := filepath.Join(dir, "corpus", name)
		if data, err := os.ReadFile(candidate); err == nil {
			var vectors []vector
			if err := json.Unmarshal(data, &vectors); err != nil {
				t.Fatalf("parsing %s: %v", candidate, err)
			}
			return vectors
		}
	}
	t.Fatal("corpus directory not found")
	return nil
}

func (v *vector) format() NumFormat {
	if v.Format == nil {
		return Invariant
	}
	return NumFormat{
		DecimalSep: []rune(v.Format.DecimalSep)[0],
		GroupSep:   []rune(v.Format.GroupSep)[0],
		Styles:     NumStyles(v.Format.Flags),
	}
}

var expectedReason = map[string]CastFailure{
	"empty":        Empty,
	"malformed":    Malformed,
	"out_of_range": OutOfRange,
}

// assertVerdict checks Go's union idiom the way callers consume it: nil fault is the
// success case, a non-nil fault carries the reason and span.
func assertVerdict[V comparable](t *testing.T, domain string, v *vector, value V, fault *Fault, expected V) {
	t.Helper()
	if fault == nil {
		if v.Expect != "ok" {
			t.Errorf("%s: %q unexpectedly parsed to %v", domain, v.Input, value)
			return
		}
		if value != expected {
			t.Errorf("%s: %q -> %v, want %v", domain, v.Input, value, expected)
		}
		return
	}
	reason, known := expectedReason[v.Expect]
	if !known {
		t.Errorf("%s: %q expected %q but faulted: %v", domain, v.Input, v.Expect, fault)
		return
	}
	if fault.Reason != reason {
		t.Errorf("%s: %q -> %v, want %v", domain, v.Input, fault.Reason, reason)
	}
	if v.FaultSpan != nil && (fault.Offset != v.FaultSpan[0] || fault.Length != v.FaultSpan[1]) {
		t.Errorf("%s: %q fault span (%d,%d), want (%d,%d)",
			domain, v.Input, fault.Offset, fault.Length, v.FaultSpan[0], v.FaultSpan[1])
	}
}

func TestBooleanCorpus(t *testing.T) {
	for _, v := range corpus(t, "boolean.json") {
		var expected bool
		if v.Value != nil {
			expected = string(v.Value) == "true"
		}
		value, fault := Bool(v.Input)
		assertVerdict(t, "boolean", &v, value, fault, expected)
	}
}

func TestIntegerCorpus(t *testing.T) {
	for _, v := range corpus(t, "integer.json") {
		format := v.format()
		switch v.Type {
		case "i8":
			value, fault := I8(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, int8(intValue(t, &v)))
		case "i16":
			value, fault := I16(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, int16(intValue(t, &v)))
		case "i32":
			value, fault := I32(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, int32(intValue(t, &v)))
		case "i64":
			value, fault := I64(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, intValue(t, &v))
		case "u8":
			value, fault := U8(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, uint8(uintValue(t, &v)))
		case "u16":
			value, fault := U16(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, uint16(uintValue(t, &v)))
		case "u32":
			value, fault := U32(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, uint32(uintValue(t, &v)))
		case "u64":
			value, fault := U64(v.Input, format)
			assertVerdict(t, "integer", &v, value, fault, uintValue(t, &v))
		default:
			t.Fatalf("integer: unknown type %q", v.Type)
		}
	}
}

func intValue(t *testing.T, v *vector) int64 {
	t.Helper()
	if v.Value == nil {
		return 0
	}
	parsed, err := strconv.ParseInt(string(v.Value), 10, 64)
	if err != nil {
		t.Fatalf("integer: %q bad expected value: %v", v.Input, err)
	}
	return parsed
}

func uintValue(t *testing.T, v *vector) uint64 {
	t.Helper()
	if v.Value == nil {
		return 0
	}
	parsed, err := strconv.ParseUint(string(v.Value), 10, 64)
	if err != nil {
		t.Fatalf("integer: %q bad expected value: %v", v.Input, err)
	}
	return parsed
}

func TestRealCorpus(t *testing.T) {
	for _, v := range corpus(t, "real.json") {
		format := v.format()
		var expected float64
		if v.Value != nil {
			parsed, err := strconv.ParseFloat(string(v.Value), 64)
			if err != nil {
				t.Fatalf("real: %q bad expected value: %v", v.Input, err)
			}
			expected = parsed
		}
		switch v.Type {
		case "f32":
			value, fault := F32(v.Input, format)
			assertVerdict(t, "real", &v, value, fault, float32(expected))
		case "f64":
			value, fault := F64(v.Input, format)
			assertVerdict(t, "real", &v, value, fault, expected)
		default:
			t.Fatalf("real: unknown type %q", v.Type)
		}
	}
}

func TestUuidCorpus(t *testing.T) {
	for _, v := range corpus(t, "uuid.json") {
		var expected uuid.UUID
		if v.Value != nil {
			var hex string
			if err := json.Unmarshal(v.Value, &hex); err != nil {
				t.Fatal(err)
			}
			parsed, err := uuid.Parse(hex)
			if err != nil {
				t.Fatalf("uuid: %q bad expected value: %v", v.Input, err)
			}
			expected = parsed
		}
		value, fault := Uuid(v.Input)
		assertVerdict(t, "uuid", &v, value, fault, expected)
	}
}

func expectedInstant(v *vector) time.Time {
	if v.Seconds == nil {
		return time.Time{}
	}
	return time.Unix(*v.Seconds, v.Nanos).UTC()
}

func TestTimestampCorpus(t *testing.T) {
	for _, v := range corpus(t, "timestamp.json") {
		value, fault := Timestamp(v.Input)
		assertVerdict(t, "timestamp", &v, value, fault, expectedInstant(&v))
	}
}

func TestUnixCorpus(t *testing.T) {
	for _, v := range corpus(t, "unix.json") {
		value, fault := Unix(v.Input, UnixPrecision(v.Precision))
		assertVerdict(t, "unix", &v, value, fault, expectedInstant(&v))
	}
}

func TestDateCorpus(t *testing.T) {
	for _, v := range corpus(t, "date.json") {
		var expected Date
		if v.Year != nil {
			expected = Date{Year: *v.Year, Month: time.Month(v.Month), Day: v.Day}
		}
		value, fault := DateOnly(v.Input)
		assertVerdict(t, "date", &v, value, fault, expected)
	}
}

func TestTimeCorpus(t *testing.T) {
	for _, v := range corpus(t, "time.json") {
		var expected time.Duration
		if v.Expect == "ok" {
			expected = time.Duration(v.Nanos)
		}
		value, fault := TimeOfDay(v.Input)
		assertVerdict(t, "time", &v, value, fault, expected)
	}
}

func TestDurationCorpus(t *testing.T) {
	for _, v := range corpus(t, "duration.json") {
		var expected Duration
		if v.Seconds != nil {
			if v.Nanos < math.MinInt32 || v.Nanos > math.MaxInt32 {
				t.Fatalf("duration: %q corpus nanos out of i32", v.Input)
			}
			expected = Duration{Seconds: *v.Seconds, Nanos: int32(v.Nanos)}
		}
		value, fault := Span(v.Input)
		assertVerdict(t, "duration", &v, value, fault, expected)
	}
}

File: php/tests/CorpusTest.php
<?php

declare(strict_types=1);

namespace HyperCast\Tests;

use DateTimeImmutable;
use HyperCast\Cast;
use HyperCast\DateOrder;
use HyperCast\CastFailure;
use HyperCast\Duration;
use HyperCast\Fault;
use HyperCast\NumFormat;
use HyperCast\Success;
use HyperCast\UnixPrecision;
use PHPUnit\Framework\TestCase;

/**
 * Replays the shared conformance corpus (corpus/*.json at the repository root) — the same
 * files every other binding's suite replays. Fault spans are byte offsets into the UTF-8
 * input, identical to PHP string byte offsets. Decoded with JSON_BIGINT_AS_STRING so
 * u64-scale expected values survive PHP's signed-64 int.
 */
final class CorpusTest extends TestCase
{
    private const EXPECTED_REASON = [
        'empty' => CastFailure::Empty,
        'malformed' => CastFailure::Malformed,
        'out_of_range' => CastFailure::OutOfRange,
    ];

    private static function corpus(string $name): array
    {
        $dir = __DIR__;
        while ($dir !== '/') {
            $candidate = $dir . '/corpus/' . $name;
            if (is_file($candidate)) {
                return json_decode(
                    file_get_contents($candidate),
                    true,
                    512,
                    JSON_THROW_ON_ERROR | JSON_BIGINT_AS_STRING
                );
            }
            $dir = \dirname($dir);
        }
        self::fail('corpus directory not found');
    }

    private static function formatOf(array $vector): NumFormat
    {
        if (!isset($vector['format'])) {
            return NumFormat::invariant();
        }
        return new NumFormat(
            $vector['format']['decimal_sep'],
            $vector['format']['group_sep'],
            $vector['format']['flags']
        );
    }

    private function assertVerdict(string $domain, array $vector, Success|Fault $verdict, mixed $expected): void
    {
        $input = $vector['input'];
        $expect = $vector['expect'];
        // The union consumption idiom in action — match over the closed pair.
        match (true) {
            $verdict instanceof Success => (function () use ($domain, $input, $expect, $verdict, $expected) {
                $this->assertSame('ok', $expect, "{$domain}: '{$input}' unexpectedly parsed");
                if ($expected instanceof DateTimeImmutable) {
                    $this->assertEquals($expected, $verdict->value, "{$domain}: '{$input}'");
                } else {
                    $this->assertEquals($expected, $verdict->value, "{$domain}: '{$input}'");
                }
            })(),
            $verdict instanceof Fault => (function () use ($domain, $input, $expect, $vector, $verdict) {
                $this->assertArrayHasKey($expect, self::EXPECTED_REASON,
                    "{$domain}: '{$input}' expected {$expect} but faulted");
                $this->assertSame(self::EXPECTED_REASON[$expect], $verdict->reason, "{$domain}: '{$input}'");
                if (isset($vector['fault'])) {
                    $this->assertSame($vector['fault'], [$verdict->offset, $verdict->length],
                        "{$domain}: '{$input}' fault span");
                }
            })(),
        };
    }

    public function testBooleanCorpus(): void
    {
        foreach (self::corpus('boolean.json') as $vector) {
            $this->assertVerdict('boolean', $vector, Cast::bool($vector['input']), $vector['value'] ?? null);
        }
    }

    public function testIntegerCorpus(): void
    {
        foreach (self::corpus('integer.json') as $vector) {
            $door = $vector['type'];
            $verdict = Cast::$door($vector['input'], self::formatOf($vector));
            $expected = $vector['value'] ?? null;
            if ($door === 'u64' && \is_string($expected)) {
                // Beyond PHP_INT_MAX the carrier is the two's-complement bit pattern —
                // compare through the unsigned renderer, the documented consumption path.
                if ($verdict instanceof Success) {
                    $this->assertSame($expected, sprintf('%u', $verdict->value),
                        "integer: '{$vector['input']}'");
                    continue;
                }
            }
            $this->assertVerdict('integer', $vector, $verdict, $expected);
        }
    }

    public function testRealCorpus(): void
    {
        foreach (self::corpus('real.json') as $vector) {
            $door = $vector['type'] === 'f32' ? 'f32' : 'f64';
            $expected = $vector['value'] ?? null;
            if ($expected !== null) {
                $expected = (float) $expected;
                if ($door === 'f32') {
                    $expected = unpack('g', pack('g', $expected))[1];
                }
            }
            $this->assertVerdict('real', $vector, Cast::$door($vector['input'], self::formatOf($vector)), $expected);
        }
    }

    public function testUuidCorpus(): void
    {
        foreach (self::corpus('uuid.json') as $vector) {
            $expected = null;
            if (isset($vector['value'])) {
                $hex = $vector['value'];
                $expected = sprintf(
                    '%s-%s-%s-%s-%s',
                    substr($hex, 0, 8),
                    substr($hex, 8, 4),
                    substr($hex, 12, 4),
                    substr($hex, 16, 4),
                    substr($hex, 20, 12)
                );
            }
            $this->assertVerdict('uuid', $vector, Cast::uuid($vector['input']), $expected);
        }
    }

    private static function expectedInstant(array $vector): ?DateTimeImmutable
    {
        if (!isset($vector['seconds'])) {
            return null;
        }
        $base = new DateTimeImmutable("@{$vector['seconds']}");
        $micros = intdiv($vector['nanos'], 1000);
        return $micros === 0 ? $base : $base->modify("+{$micros} microseconds");
    }

    public function testTimestampCorpus(): void
    {
        foreach (self::corpus('timestamp.json') as $vector) {
            $this->assertVerdict('timestamp', $vector, Cast::timestamp($vector['input']),
                self::expectedInstant($vector));
        }
    }

    public function testUnixCorpus(): void
    {
        foreach (self::corpus('unix.json') as $vector) {
            $verdict = Cast::unix($vector['input'], UnixPrecision::from($vector['precision']));
            $this->assertVerdict('unix', $vector, $verdict, self::expectedInstant($vector));
        }
    }

    public function testDateCorpus(): void
    {
        foreach (self::corpus('date.json') as $vector) {
            $expected = null;
            if (isset($vector['year'])) {
                $expected = new DateTimeImmutable(
                    sprintf('%04d-%02d-%02d 00:00:00', $vector['year'], $vector['month'], $vector['day']),
                    new \DateTimeZone('UTC')
                );
            }
            $this->assertVerdict('date', $vector, Cast::date($vector['input']), $expected);
        }
    }

    public function testDateOrderCorpus(): void
    {
        foreach (self::corpus('date_order.json') as $vector) {
            $order = DateOrder::from($vector['order']);
            $expected = null;
            if (isset($vector['year'])) {
                $expected = new DateTimeImmutable(
                    sprintf('%04d-%02d-%02d 00:00:00', $vector['year'], $vector['month'], $vector['day']),
                    new \DateTimeZone('UTC')
                );
            }
            $this->assertVerdict('date_order', $vector, Cast::date($vector['input'], $order), $expected);
        }
    }

    public function testDateTimeCorpus(): void
    {
        foreach (self::corpus('datetime.json') as $vector) {
            $order = DateOrder::from($vector['order']);
            $expected = null;
            if (isset($vector['year'])) {
                $secondOfDay = intdiv($vector['nanos_of_day'], 1_000_000_000);
                $micros = intdiv($vector['nanos_of_day'] % 1_000_000_000, 1000);
                $expected = new DateTimeImmutable(
                    sprintf(
                        '%04d-%02d-%02d %02d:%02d:%02d.%06d',
                        $vector['year'],
                        $vector['month'],
                        $vector['day'],
                        intdiv($secondOfDay, 3600),
                        intdiv($secondOfDay % 3600, 60),
                        $secondOfDay % 60,
                        $micros
                    ),
                    new \DateTimeZone('UTC')
                );
            }
            $this->assertVerdict('datetime', $vector, Cast::datetime($vector['input'], $order), $expected);
        }
    }

    public function testTimeCorpus(): void
    {
        foreach (self::corpus('time.json') as $vector) {
            $this->assertVerdict('time', $vector, Cast::time($vector['input']), $vector['nanos'] ?? null);
        }
    }

    public function testDurationCorpus(): void
    {
        foreach (self::corpus('duration.json') as $vector) {
            $expected = isset($vector['seconds'])
                ? new Duration($vector['seconds'], $vector['nanos'])
                : null;
            $this->assertVerdict('duration', $vector, Cast::duration($vector['input']), $expected);
        }
    }
}
